odd-platform-api/src/main/java/com/provectus/oddplatform/repository/DatasetStructureRepositoryImpl.java

```diff
@@ -0,0 +1,76 @@
+package com.provectus.oddplatform.repository;
+
+import com.provectus.oddplatform.model.tables.pojos.DatasetFieldPojo;
+import com.provectus.oddplatform.model.tables.pojos.DatasetStructurePojo;
+import com.provectus.oddplatform.model.tables.pojos.DatasetVersionPojo;
+import com.provectus.oddplatform.model.tables.records.DatasetStructureRecord;
+import lombok.RequiredArgsConstructor;
+import org.jooq.DSLContext;
+import org.jooq.InsertSetStep;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.provectus.oddplatform.model.Tables.DATASET_STRUCTURE;
+
+@Repository
+@RequiredArgsConstructor
+public class DatasetStructureRepositoryImpl implements DatasetStructureRepository {
+    private final DSLContext dslContext;
+    private final DatasetFieldRepository datasetFieldRepository;
+    private final DatasetVersionRepository datasetVersionRepository;
+
+    @Override
+    @Transactional
+    public List<DatasetStructurePojo> bulkCreate(final List<DatasetVersionPojo> versions,
+                                                 final Map<Long, List<DatasetFieldPojo>> datasetFields) {
+        final List<DatasetFieldPojo> fields = datasetFields.values().stream()
+            .flatMap(List::stream)
+            .collect(Collectors.toList());
+
+        final List<DatasetVersionPojo> createdVersions = datasetVersionRepository.bulkCreate(versions);
+
+        final Map<String, DatasetFieldPojo> createdDFMap = datasetFieldRepository.bulkCreateIfNotExist(fields)
+            .stream()
+            .collect(Collectors.toMap(DatasetFieldPojo::getOddrn, Function.identity()));
+
+        final List<DatasetStructurePojo> structure = createdVersions.stream()
+            .flatMap(v -> datasetFields.get(v.getDatasetId()).stream()
+                .map(f -> createdDFMap.get(f.getOddrn()))
+                .map(DatasetFieldPojo::getId)
+                .map(dfId -> new DatasetStructurePojo().setDatasetFieldId(dfId).setDatasetVersionId(v.getId())))
+            .collect(Collectors.toList());
+
+        return bulkInsert(structure);
+    }
+
+    protected List<DatasetStructurePojo> bulkInsert(final List<DatasetStructurePojo> entities) {
+        if (entities.isEmpty()) {
+            return Collections.emptyList();
```

Comment:
I find this behaviour kinda misleading.
Some places (e.g. DatasetFieldRepositoryImpl) we do return concatenated list of both newly created and previously existed records in this list and in some places like this we do return an empty collection.

Reply:
Didn't quite understand your point. In both places we don't wanna start building a query in case when no data is present. Can you elaborate?

Reply:
I'm considering consistence. kinda concerned about different behaviour in different places.

Reply:
Like, I wouldn't know will it return an empty collection or a collection with all the items (including the ones already existing) without checking the whole chain of underlying methods.